// bucklescript-tea/src-ocaml/tea_svg.js
// Generated by BUCKLESCRIPT VERSION 4.0.6, PLEASE EDIT WITH CARE

import * as Vdom from "./vdom.js";
import * as Block from "bs-platform/lib/es6/block.js";

var svgNamespace = "http://www.w3.org/2000/svg";

function text(str) {
  return /* Text */Block.__(1, [str]);
}

var lazy1 = Vdom.lazyGen;

function node(tagName, $staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, tagName, key, unique, props, nodes);
}

function svg($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "svg", key, unique, props, nodes);
}

function foreignObject($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "foreignObject", key, unique, props, nodes);
}

function animate($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "animate", key, unique, props, nodes);
}

function animateColor($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "animateColor", key, unique, props, nodes);
}

function animateMotion($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "animateMotion", key, unique, props, nodes);
}

function animateTransform($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "animateTransform", key, unique, props, nodes);
}

function mpath($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "mpath", key, unique, props, nodes);
}

function set($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "set", key, unique, props, nodes);
}

function a($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "a", key, unique, props, nodes);
}

function defs($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "defs", key, unique, props, nodes);
}

function g($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "g", key, unique, props, nodes);
}

function marker($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "marker", key, unique, props, nodes);
}

function mask($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "mask", key, unique, props, nodes);
}

function missingGlyph($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "missingGlyph", key, unique, props, nodes);
}

function pattern($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "pattern", key, unique, props, nodes);
}

function $$switch($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "switch", key, unique, props, nodes);
}

function symbol($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "symbol", key, unique, props, nodes);
}

function desc($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "desc", key, unique, props, nodes);
}

function metadata($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "metadata", key, unique, props, nodes);
}

function title($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "title", key, unique, props, nodes);
}

function feBlend($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feBlend", key, unique, props, nodes);
}

function feColorMatrix($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feColorMatrix", key, unique, props, nodes);
}

function feComponentTransfer($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feComponentTransfer", key, unique, props, nodes);
}

function feComposite($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feComposite", key, unique, props, nodes);
}

function feConvolveMatrix($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feConvolveMatrix", key, unique, props, nodes);
}

function feDiffuseLighting($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feDiffuseLighting", key, unique, props, nodes);
}

function feDisplacementMap($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feDisplacementMap", key, unique, props, nodes);
}

function feFlood($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feFlood", key, unique, props, nodes);
}

function feFuncA($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feFuncA", key, unique, props, nodes);
}

function feFuncB($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feFuncB", key, unique, props, nodes);
}

function feFuncG($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feFuncG", key, unique, props, nodes);
}

function feFuncR($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feFuncR", key, unique, props, nodes);
}

function feGaussianBlur($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feGaussianBlur", key, unique, props, nodes);
}

function feImage($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feImage", key, unique, props, nodes);
}

function feMerge($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feMerge", key, unique, props, nodes);
}

function feMergeNode($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feMergeNode", key, unique, props, nodes);
}

function feMorphology($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feMorphology", key, unique, props, nodes);
}

function feOffset($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feOffset", key, unique, props, nodes);
}

function feSpecularLighting($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feSpecularLighting", key, unique, props, nodes);
}

function feTile($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feTile", key, unique, props, nodes);
}

function feTurbulence($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feTurbulence", key, unique, props, nodes);
}

function font($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "font", key, unique, props, nodes);
}

function fontFace($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "fontFace", key, unique, props, nodes);
}

function fontFaceFormat($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "fontFaceFormat", key, unique, props, nodes);
}

function fontFaceName($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "fontFaceName", key, unique, props, nodes);
}

function fontFaceSrc($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "fontFaceSrc", key, unique, props, nodes);
}

function fontFaceUri($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "fontFaceUri", key, unique, props, nodes);
}

function hkern($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "hkern", key, unique, props, nodes);
}

function vkern($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "vkern", key, unique, props, nodes);
}

function linearGradient($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "linearGradient", key, unique, props, nodes);
}

function radialGradient($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "radialGradient", key, unique, props, nodes);
}

function stop($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "stop", key, unique, props, nodes);
}

function circle($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "circle", key, unique, props, nodes);
}

function ellipse($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "ellipse", key, unique, props, nodes);
}

function svgimage($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "image", key, unique, props, nodes);
}

function line($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "line", key, unique, props, nodes);
}

function path($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "path", key, unique, props, nodes);
}

function polygon($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "polygon", key, unique, props, nodes);
}

function polyline($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "polyline", key, unique, props, nodes);
}

function rect($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "rect", key, unique, props, nodes);
}

function use($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "use", key, unique, props, nodes);
}

function feDistantLight($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feDistantLight", key, unique, props, nodes);
}

function fePointLight($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "fePointLight", key, unique, props, nodes);
}

function feSpotLight($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "feSpotLight", key, unique, props, nodes);
}

function altGlyph($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "altGlyph", key, unique, props, nodes);
}

function altGlyphDef($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "altGlyphDef", key, unique, props, nodes);
}

function altGlyphItem($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "altGlyphItem", key, unique, props, nodes);
}

function glyph($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "glyph", key, unique, props, nodes);
}

function glyphRef($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "glyphRef", key, unique, props, nodes);
}

function textPath($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "textPath", key, unique, props, nodes);
}

function text$prime($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "text", key, unique, props, nodes);
}

function tref($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "tref", key, unique, props, nodes);
}

function tspan($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "tspan", key, unique, props, nodes);
}

function clipPath($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "clipPath", key, unique, props, nodes);
}

function svgcolorProfile($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "colorProfile", key, unique, props, nodes);
}

function cursor($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "cursor", key, unique, props, nodes);
}

function filter($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "filter", key, unique, props, nodes);
}

function script($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "script", key, unique, props, nodes);
}

function style($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "style", key, unique, props, nodes);
}

function view($staropt$star, $staropt$star$1, props, nodes) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  var unique = $staropt$star$1 !== undefined ? $staropt$star$1 : "";
  return Vdom.fullnode(svgNamespace, "view", key, unique, props, nodes);
}

var Cmds = 0;

var Attributes = 0;

var Events = 0;

var noNode = Vdom.noNode;

export {
  Cmds ,
  Attributes ,
  Events ,
  svgNamespace ,
  noNode ,
  text ,
  lazy1 ,
  node ,
  svg ,
  foreignObject ,
  animate ,
  animateColor ,
  animateMotion ,
  animateTransform ,
  mpath ,
  set ,
  a ,
  defs ,
  g ,
  marker ,
  mask ,
  missingGlyph ,
  pattern ,
  $$switch ,
  symbol ,
  desc ,
  metadata ,
  title ,
  feBlend ,
  feColorMatrix ,
  feComponentTransfer ,
  feComposite ,
  feConvolveMatrix ,
  feDiffuseLighting ,
  feDisplacementMap ,
  feFlood ,
  feFuncA ,
  feFuncB ,
  feFuncG ,
  feFuncR ,
  feGaussianBlur ,
  feImage ,
  feMerge ,
  feMergeNode ,
  feMorphology ,
  feOffset ,
  feSpecularLighting ,
  feTile ,
  feTurbulence ,
  font ,
  fontFace ,
  fontFaceFormat ,
  fontFaceName ,
  fontFaceSrc ,
  fontFaceUri ,
  hkern ,
  vkern ,
  linearGradient ,
  radialGradient ,
  stop ,
  circle ,
  ellipse ,
  svgimage ,
  line ,
  path ,
  polygon ,
  polyline ,
  rect ,
  use ,
  feDistantLight ,
  fePointLight ,
  feSpotLight ,
  altGlyph ,
  altGlyphDef ,
  altGlyphItem ,
  glyph ,
  glyphRef ,
  textPath ,
  text$prime ,
  tref ,
  tspan ,
  clipPath ,
  svgcolorProfile ,
  cursor ,
  filter ,
  script ,
  style ,
  view ,
  
}
/* No side effect */
