// bucklescript-tea/src-ocaml/tea_mouse.js
// Generated by BUCKLESCRIPT VERSION 4.0.6, PLEASE EDIT WITH CARE

import * as Vdom from "./vdom.js";
import * as Block from "bs-platform/lib/es6/block.js";
import * as Curry from "bs-platform/lib/es6/curry.js";
import * as Tea_sub from "./tea_sub.js";
import * as Tea_json from "./tea_json.js";
import * as Js_primitive from "bs-platform/lib/es6/js_primitive.js";

var position = Tea_json.Decoder[/* map2 */17]((function (x, y) {
        return /* record */[
                /* x */x,
                /* y */y
              ];
      }), Tea_json.Decoder[/* field */11]("pageX", Tea_json.Decoder[/* int */3]), Tea_json.Decoder[/* field */11]("pageY", Tea_json.Decoder[/* int */3]));

function registerGlobal(name, key, tagger) {
  var enableCall = function (callbacks_base) {
    var callbacks = /* record */[/* contents */callbacks_base];
    var fn = function (ev) {
      var match = Tea_json.Decoder[/* decodeEvent */31](position, ev);
      if (match.tag) {
        return undefined;
      } else {
        return Js_primitive.some(Curry._1(tagger, match[0]));
      }
    };
    var handler = /* EventHandlerCallback */Block.__(0, [
        key,
        fn
      ]);
    var elem = document;
    var cache = Vdom.eventHandler_Register(callbacks, elem, name, handler);
    return (function () {
        Vdom.eventHandler_Unregister(elem, name, cache);
        return /* () */0;
      });
  };
  return Tea_sub.registration(key, enableCall);
}

function clicks($staropt$star, tagger) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  return registerGlobal("click", key, tagger);
}

function moves($staropt$star, tagger) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  return registerGlobal("mousemove", key, tagger);
}

function downs($staropt$star, tagger) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  return registerGlobal("mousedown", key, tagger);
}

function ups($staropt$star, tagger) {
  var key = $staropt$star !== undefined ? $staropt$star : "";
  return registerGlobal("mouseup", key, tagger);
}

export {
  position ,
  registerGlobal ,
  clicks ,
  moves ,
  downs ,
  ups ,
  
}
/* position Not a pure module */
