// bucklescript-tea/src-ocaml/web_window.js
// Generated by BUCKLESCRIPT VERSION 4.0.6, PLEASE EDIT WITH CARE


function history() {
  return window.history;
}

function localStorage() {
  return window.localStorage;
}

function $$location() {
  return window.location;
}

function requestAnimationFrame(callback) {
  return window.requestAnimationFrame(callback);
}

function cancelAnimationFrame(id) {
  return window.cancelAnimationFrame(id);
}

function $$clearTimeout(id) {
  return window.clearTimeout(id);
}

function $$setInterval(cb, msTime) {
  return window.setInterval(cb, msTime);
}

function $$setTimeout(cb, msTime) {
  return window.setTimeout(cb, msTime);
}

function addEventListener(typ, listener, options) {
  return window.addEventListener(typ, listener, options);
}

function removeEventListener(typ, listener, options) {
  return window.removeEventListener(typ, listener, options);
}

function requestAnimationFrame_polyfill() {
  return (
  // requestAnimationFrame polyfill
  (function() {
      var lastTime = 0;
      var vendors = ['ms', 'moz', 'webkit', 'o'];
      for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
          window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
          window.cancelAnimationFrame = window[vendors[x]+'CancelAnimationFrame']
                                     || window[vendors[x]+'CancelRequestAnimationFrame'];
      }

      if (!window.requestAnimationFrame)
          window.requestAnimationFrame = function(callback, element) {
              var currTime = new Date().getTime();
              var timeToCall = Math.max(0, 16 - (currTime - lastTime));
              var id = window.setTimeout(function() { callback(currTime + timeToCall); },
                timeToCall);
              lastTime = currTime + timeToCall;
              return id;
          };

      if (!window.cancelAnimationFrame)
          window.cancelAnimationFrame = function(id) {
              clearTimeout(id);
          };
  }())
  );
}

var History = 0;

var LocalStorage = 0;

export {
  History ,
  LocalStorage ,
  history ,
  localStorage ,
  $$location ,
  requestAnimationFrame ,
  cancelAnimationFrame ,
  $$clearTimeout ,
  $$setInterval ,
  $$setTimeout ,
  addEventListener ,
  removeEventListener ,
  requestAnimationFrame_polyfill ,
  
}
/* No side effect */
