// bucklescript-tea/src-ocaml/web_location.js
// Generated by BUCKLESCRIPT VERSION 4.0.6, PLEASE EDIT WITH CARE


function getHref($$location) {
  return $$location.href;
}

function setHref($$location, value) {
  $$location.href = value;
  return /* () */0;
}

function getProtocol($$location) {
  return $$location.protocol;
}

function setProtocol($$location, value) {
  $$location.protocol = value;
  return /* () */0;
}

function getHost($$location) {
  return $$location.host;
}

function setHost($$location, value) {
  $$location.host = value;
  return /* () */0;
}

function getHostname($$location) {
  return $$location.hostname;
}

function setHostname($$location, value) {
  $$location.hostname = value;
  return /* () */0;
}

function getPort($$location) {
  return $$location.port;
}

function setPort($$location, value) {
  $$location.port = value;
  return /* () */0;
}

function getPathname($$location) {
  return $$location.pathname;
}

function setPathname($$location, value) {
  $$location.pathname = value;
  return /* () */0;
}

function getSearch($$location) {
  return $$location.search;
}

function setSearch($$location, value) {
  $$location.search = value;
  return /* () */0;
}

function getHash($$location) {
  return $$location.hash;
}

function setHash($$location, value) {
  $$location.hash = value;
  return /* () */0;
}

function getUsername($$location) {
  return $$location.username;
}

function setUsername($$location, value) {
  $$location.username = value;
  return /* () */0;
}

function getPassword($$location) {
  return $$location.password;
}

function setPassword($$location, value) {
  $$location.password = value;
  return /* () */0;
}

function getOrigin($$location) {
  return $$location.origin;
}

function asRecord($$location) {
  return /* record */[
          /* href */$$location.href,
          /* protocol */$$location.protocol,
          /* host */$$location.host,
          /* hostname */$$location.hostname,
          /* port */$$location.port,
          /* pathname */$$location.pathname,
          /* search */$$location.search,
          /* hash */$$location.hash,
          /* username */$$location.username,
          /* password */$$location.password,
          /* origin */$$location.origin
        ];
}

export {
  getHref ,
  setHref ,
  getProtocol ,
  setProtocol ,
  getHost ,
  setHost ,
  getHostname ,
  setHostname ,
  getPort ,
  setPort ,
  getPathname ,
  setPathname ,
  getSearch ,
  setSearch ,
  getHash ,
  setHash ,
  getUsername ,
  setUsername ,
  getPassword ,
  setPassword ,
  getOrigin ,
  asRecord ,
  
}
/* No side effect */
